text/2020-09-28-memory-trace.md

```diff
@@ -0,0 +1,131 @@
+# Fully instrumented memory trace
+
+## Summary
+
+Make TiKV fully instrumented to provide accurate and details memory trace.
+
+## Motivation
+
+Suspicious memory increase has been a big challenge for TiKV. We don't have a
+good tool to know who is exhausted the memory. With the help of profile tools,
+we can identify memory leaks that can be reproduced and grow rapidly. However,
+using profiling tools requires a recompilation and also has a performance
+impact. And profile tool can only provide information about allocations, for
+diagnostics, find out who is holding the memory is more useful.
+
+Another use case is to limit the memory usage of a query. This is very useful
```

Comment:
How would we use this to limit the memory usage of a query? It seems the proposal is to take occasional snapshots. So if a query held memory for long enough to show up in a snapshot we would limit it? What would the snapshot frequency be if we were to use this proactively rather than just for debug? What would the overhead be?

Reply:
A query can check the current memory usage after scanning constant row count. The cost should be small enough as the memory needs to be collected can be calculated for every result row, which is just query the encoded byte array's length. And the memory usage should be able to be caught in time because memory usage for a row is tiny generally.

Reply:
This seems like it might be implemented separately from this proposal? Or is the query asking for a snapshot of its own usage after every row count?

Reply:
Stats collection is implemented in this proposal, but query and abort should be implemented in other proposal like QoS.